opus/application/test_api/api_test_helper.py

```diff
@@ -137,13 +140,23 @@ def _run_html_equal_file(self, url, exp_file):
             expected = fp.read()
         expected = self._clean_string(expected)
         resp = self._clean_string(str(response.content))
+        if embedded_dynamic_image:
+            # extract the dynamic images and replace with generic tests.
+            expected, expected_images = self.__extract_images(expected)
+            resp, resp_images = self.__extract_images(resp)
+        else:
+            expected_images = resp_images = []
         print('Got:')
         print(resp)
         print('Expected:')
         print(expected)
         if resp != expected:
             self._print_clean_diffs(resp, expected)
```

Comment:
This prints the diffs even if the images end up being the same. It seems like we wouldn't want to do that.

Reply:
A couple of questions. 

Some projects have a way of specifying requirements needed only for testing. Does this have that? If yes, I'll add numpy there. If not, I'll get rid of numpy and compare the two images bit by bit. Slower, but fast enough for a test. 

The print statements have always been there. I didn't add them. But you only see the output of the print statements in the log when the test fails. They turned out to be incredibly useful for me.

Reply:
Because OPUS isn't designed as an installable module, there's currently only one set of requirements. In our other projects that you can do "pip install" on, pyproject.toml has the installation requirements and requirements.txt has all the requirements including testing, doc generation, linting, etc.

Regarding the print statements, I meant the diffs when the strings differ. Normally if the strings differ, the test fails. In this case, if the strings differ, the test can still pass. I thought we should maybe only print the differences if the test is going to fail. But it's not a big deal.

Reply:
I will also add.  This print("got").... sequence of code appears 11 times in the file. pytest only shows stdout if the test fails, so I'm assuming it's harmless.